presto-spi/src/main/java/com/facebook/presto/spi/PageBuilder.java

```diff
@@ -77,14 +78,24 @@ private PageBuilder(int initialExpectedEntries, int maxPageBytes, List<? extends
             }
         }
 
-        blockBuilders = new BlockBuilder[types.size()];
-        for (int i = 0; i < blockBuilders.length; i++) {
-            blockBuilders[i] = types.get(i).createBlockBuilder(
-                    pageBuilderStatus.createBlockBuilderStatus(),
-                    expectedEntries,
-                    pageBuilderStatus.getMaxBlockSizeInBytes() / expectedEntries);
-        }
         this.initialExpectedEntries = expectedEntries;
+
+        if (templateBlockBuilders.isPresent()) {
+            BlockBuilder[] templates = templateBlockBuilders.get();
+            blockBuilders = new BlockBuilder[templates.length];
+            for (int i = 0; i < blockBuilders.length; i++) {
+                blockBuilders[i] = templates[i].newBlockBuilderLike(pageBuilderStatus.createBlockBuilderStatus());
+            }
+        }
+        else {
+            blockBuilders = new BlockBuilder[types.size()];
```

Comment:
the assignment to `blockBuilders` can be moved out of the if/else (for both branches)